tests/posix/pthread_pressure/src/main.c

```diff
@@ -0,0 +1,230 @@
+
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/ztest.h>
+
+#define STACK_SIZE K_THREAD_STACK_LEN(CONFIG_TEST_STACK_SIZE)
+
+/* update interval for printing stats */
+#if CONFIG_TEST_DURATION_S < 10
+#define UPDATE_INTERVAL_S 1
+#elif CONFIG_TEST_DURATION_S < 30
+#define UPDATE_INTERVAL_S 5
+#else
+#define UPDATE_INTERVAL_S 10
+#endif
+
+/* 32 threads is mainly a limitation of find_lsb_set() */
+#define NUM_THREADS MIN(32, MIN(CONFIG_TEST_NUM_CPUS, CONFIG_MAX_PTHREAD_COUNT))
+
+typedef int (*create_fn)(int i);
+typedef int (*join_fn)(int i);
+
+static void *setup(void);
+static void before(void *fixture);
+
+static LOG_MODULE_REGISTER(pthread_pressure, LOG_LEVEL_INF);
+
+/* bitmask of available threads */
+static atomic_t free_threads = ATOMIC_INIT(BIT_MASK(NUM_THREADS));
+
+/* array of thread stacks */
+static K_THREAD_STACK_ARRAY_DEFINE(thread_stacks, NUM_THREADS, STACK_SIZE);
+
+static struct k_thread k_threads[NUM_THREADS];
+static size_t counters[NUM_THREADS];
+
+static void print_stats(uint64_t now, uint64_t end)
+{
+	LOG_INF("now: %" PRIu64 " end: %" PRIu64 " (ticks)", now, end);
+
+	for (int i = 0; i < NUM_THREADS; ++i) {
+		LOG_INF("Thread %d created and joined %zu times", i, counters[i]);
+	}
+}
+
+static void test_create_join_common(const char *tag, create_fn create, join_fn join)
+{
+	int i;
+	int ret;
+	uint64_t now_ticks;
+	uint32_t maybe_free_threads;
+	const uint64_t end_ticks =
+		sys_clock_timeout_end_calc(K_MSEC(MSEC_PER_SEC * CONFIG_TEST_DURATION_S));
+	uint64_t update_ticks =
+		sys_clock_timeout_end_calc(K_MSEC(MSEC_PER_SEC * UPDATE_INTERVAL_S));
+
+	LOG_INF("NUM_THREADS: %u", NUM_THREADS);
+	LOG_INF("TEST_NUM_CPUS: %u", CONFIG_TEST_NUM_CPUS);
+	LOG_INF("TEST_DURATION_S: %u", CONFIG_TEST_DURATION_S);
+	LOG_INF("TEST_DELAY_US: %u", CONFIG_TEST_DELAY_US);
+
+	for (i = 0; i < NUM_THREADS; ++i) {
+		/* mark thread i as in-use */
+		atomic_clear_bit(&free_threads, i);
+
+		/* spawn thread i */
+		ret = create(i);
+		zassert_ok(ret, "%s_create(%d)[%zu] failed: %d", i, counters[i], ret);
+		LOG_DBG("spawned thread %d", i);
+	}
+
+	do {
+		/* allow the test thread to be swapped-out, utilizing all cores */
+		k_yield();
+
+		/* "maybe": window after bit is set before terminating and becoming joinable */
+		maybe_free_threads = (uint32_t)atomic_get(&free_threads);
+		LOG_DBG("free_threads: %x", maybe_free_threads);
+
+		while (maybe_free_threads != 0) {
+
+			/* shouldn't matter if we join 0 or N-1 first */
+			i = find_lsb_set(maybe_free_threads) - 1;
+			maybe_free_threads &= ~BIT(i);
+
+			/* join thread i */
+			LOG_DBG("joining thread %d", i);
+			ret = join(i);
+			zassert_ok(ret, "%s_join(%d)[%zu] failed: %d", i, counters[i], ret);
+
+			/* update counter i after each (create,join) pair */
+			++counters[i];
+
+			/* success with 0 delay means we are ~raceless */
+			k_busy_wait(CONFIG_TEST_DELAY_US);
+
+			/* re-spawn thread i */
+			ret = create(i);
```

Comment:
Probably I'm missing something in the test logic, but shouldn't we clear thread bit in `free_threads` here?
```suggestion
			/* re-spawn thread i */
			atomic_clear_bit(&free_threads, i);
			ret = create(i);
```

Otherwise all bits in `free_threads` will be set after the fist iteration of thread create + thread join.

Reply:
That sounds about right.

Reply:
@cfriedt 

Though (looking a bit more in details into the test), probably it's not a big issue right now (or even the corner case we accidentally bumped into)

Due to this 'bug' we might join the thread which hasn't started yet (or which is starting right now on another CPU core) in this test.

But, as I understan, both options are vaild - so probably it's worth to keep this 'bug'

Reply:
It unfortunately makes no difference and the race condition still exists.